File: core/src/main/java/com/palomiklo/streampulse/context/AsynchronousContext.java
package com.palomiklo.streampulse.context;

import jakarta.servlet.AsyncContext;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

public record AsynchronousContext(HttpServletRequest request, AsyncContext asyncContext) {
    public static AsynchronousContext startAsyncContext(HttpServletRequest request, HttpServletResponse response) {
        request.setAttribute("org.apache.catalina.ASYNC_SUPPORTED", true);
        AsyncContext asyncContext = request.startAsync(request, response);
        asyncContext.setTimeout(0);
        return new AsynchronousContext(request, asyncContext);
    }
}

File: core/src/main/java/com/palomiklo/streampulse/context/AsynchronousContext.kt
package com.palomiklo.streampulse.context

import jakarta.servlet.AsyncContext
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse

@JvmRecord
data class AsynchronousContext(val req: HttpServletRequest, val actx: AsyncContext) {
    companion object {
        fun startAsyncContext(req: HttpServletRequest, res: HttpServletResponse?): AsynchronousContext {
            req.setAttribute("org.apache.catalina.ASYNC_SUPPORTED", true)
            val actx = req.startAsync(req, res)
            actx.timeout = 0
            return AsynchronousContext(req, actx)
        }
    }
}
